subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/AttributeConfigurationSelector.java

```diff
@@ -53,37 +55,47 @@ public static VariantSelectionResult selectVariantsUsingAttributeMatching(Immuta
 
     private static VariantSelectionResult selectVariantsUsingAttributeMatching(ImmutableAttributes consumerAttributes, Collection<? extends Capability> explicitRequestedCapabilities, ComponentGraphResolveState targetComponentState, AttributesSchemaInternal consumerSchema, List<IvyArtifactName> requestedArtifacts, AttributeMatchingExplanationBuilder explanationBuilder) {
         ComponentGraphResolveMetadata targetComponent = targetComponentState.getMetadata();
+        AttributeMatcher attributeMatcher = consumerSchema.withProducer(targetComponent.getAttributesSchema());
+
         Optional<List<? extends VariantGraphResolveMetadata>> variantsForGraphTraversal = targetComponent.getVariantsForGraphTraversal();
-        List<? extends VariantGraphResolveMetadata> consumableVariants = variantsForGraphTraversal.or(ImmutableList.of());
-        AttributesSchemaInternal producerAttributeSchema = targetComponent.getAttributesSchema();
-        AttributeMatcher attributeMatcher = consumerSchema.withProducer(producerAttributeSchema);
-        ConfigurationGraphResolveMetadata fallbackConfiguration = targetComponent.getConfiguration(Dependency.DEFAULT_CONFIGURATION);
-        if (fallbackConfiguration != null && !fallbackConfiguration.isCanBeConsumed()) {
-            fallbackConfiguration = null;
-        }
-        ModuleVersionIdentifier versionId = targetComponent.getModuleVersionId();
-        if (!consumableVariants.isEmpty()) {
-            ImmutableList<VariantGraphResolveMetadata> variantsProvidingRequestedCapabilities = filterVariantsByRequestedCapabilities(targetComponent, explicitRequestedCapabilities, consumableVariants, versionId.getGroup(), versionId.getName(), true);
-            if (variantsProvidingRequestedCapabilities.isEmpty()) {
-                throw new NoMatchingCapabilitiesException(targetComponent, explicitRequestedCapabilities, consumableVariants);
+        boolean variantAware = variantsForGraphTraversal.isPresent();
+
+        // Fallback to the default configuration if there are no variants or if variant aware resolution is not supported.
+        if (!variantAware || variantsForGraphTraversal.get().isEmpty()) {
```

Comment:
💭 The code around here seems like it ought to be extracted to a separate `attemptToMatchDefault()` method.